### PR TITLE
External window

### PR DESCRIPTION
**Pull request by Meakk**

### CODE REVIEW

File: src/library/f3d_window_impl.h [hunk 1]
```diff
@@ -21,6 +21,7 @@ class window_impl : public window
     const std::string& outputFile = "") override;
   bool setIcon(const void* icon, size_t iconSize) override;
   bool setWindowName(const std::string& windowName) override;
+  bool setSize(int width, int height) override;
```
Comment (mwestphal) [inline src/library/f3d_window_impl.h:24]: resolution is already in f3d::options

File: library/private/f3d_window_impl_standard.h [hunk 2]
```diff
@@ -18,7 +18,14 @@ namespace detail
...
 public:
-  window_impl_standard(const options& options, bool offscreen);
+  enum class WindowType : unsigned char
+  {
+    NATIVE,
+    NATIVE_OFFSCREEN,
+    EXTERNAL
+  };
+
+  window_impl_standard(const options& options, WindowType type);
```
Comment (mwestphal) [inline library/private/f3d_window_impl_standard.h:28]: maybe use the engine flags directy ?

Comment (Meakk) [reply library/private/f3d_window_impl_standard.h]: I don't know, adding `#include "engine.h"` just to hack an existing enum does not look clean to me, I prefer a specific and private enum.

File: library/src/engine.cxx [hunk 3]
```diff
@@ -34,8 +34,20 @@ engine::engine(const flags_t& flags)
...
   if (flags & CREATE_WINDOW)
   {
-    this->Internals->Window = std::make_unique<detail::window_impl_standard>(
-      *this->Internals->Options, flags & WINDOW_OFFSCREEN);
+    detail::window_impl_standard::WindowType type =
+      detail::window_impl_standard::WindowType::NATIVE;
+
+    if (flags & WINDOW_EXTERNAL)
+    {
+      type = detail::window_impl_standard::WindowType::EXTERNAL;
```
Comment (Meakk) [inline library/src/engine.cxx:42]: @mwestphal any idea why it's not covered by `TestSDKExternalWindowQT.cxx`?

Comment (Meakk) [reply library/src/engine.cxx:42]: Nevermind, I know why, I need to update the Docker image

File: library/public/engine.h [hunk 4]
```diff
@@ -51,14 +51,16 @@ class F3D_EXPORT engine
    * engine::CREATE_WINDOW: Create a window to render into.
    * engine::CREATE_INTERACTOR: Create an interactor to interact with.
    * engine::WINDOW_OFFSCREEN: Create an offscreen window to render into, need CREATE_WINDOW.
+   * engine::WINDOW_EXTERNAL: Use an external window to render into, need CREATE_WINDOW.
```
Comment (mwestphal) [inline library/public/engine.h:54]: details which flags has precendence

File: library/testing/TestSDKExternalWindowQT.cxx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    QImage frameBuffer = grabFramebuffer().mirrored().convertToFormat(QImage::Format_RGB888);
+
+    f3d::image img;
+    img.setResolution(frameBuffer.width(), frameBuffer.height())
+      .setChannelCount(3)
+      .setData(frameBuffer.bits());
+
+    f3d::image diff;
+    double error;
+    if (!img.compare(f3d::image(mBaseline), diff, 50, error))
```
Comment (mwestphal) [inline library/testing/TestSDKExternalWindowQT.cxx:50]: use TestSDKHelpers ?

File: library/testing/TestSDKExternalWindowGLFW.cxx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  while (!glfwWindowShouldClose(window) && glfwGetTime() < 1.0)
+  {
+    eng.getWindow().render();
+    glfwSwapBuffers(window);
+    glfwPollEvents();
+  }
+
+  glfwTerminate();
+
+  return EXIT_SUCCESS;
```
[the remark this answers is not in the document]
Comment (mwestphal) [reply library/testing/TestSDKExternalWindowGLFW.cxx:71]: then comparing using window.renderToImage should be an ok alternative

File: library/public/image.h [hunk 7]
```diff
@@ -100,7 +100,7 @@ class F3D_EXPORT image
    * Save an image to a file in .png format
    * Can throw an image::exception if image is not sane.
    */
-  image& save(const std::string& path);
+  const image& save(const std::string& path) const;
```
Comment (mwestphal) [inline library/public/image.h:103]: I think a global constness pass on the API should be done

Comment (mwestphal) [reply library/public/image.h:103]: out of scope ofc

Comment (Meakk) [reply library/public/image.h:103]: I agree

File: library/testing/TestSDKExternalWindowGLFW.cxx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    });
+
+  while (!glfwWindowShouldClose(window) && glfwGetTime() < 1.0)
+  {
+    eng.getWindow().render();
+    glfwSwapBuffers(window);
+    glfwPollEvents();
+  }
+
+  if (!TestSDKHelpers::RenderTest(
```
Comment (mwestphal) [inline library/testing/TestSDKExternalWindowGLFW.cxx:71]: Add a comment about the limitation of this image test

**Review by mwestphal — APPROVED** [review]

**Pull request merged**